src/ip_resolver.cpp

```diff
@@ -20,6 +20,73 @@
 
 #include "ip_resolver.hpp"
 
+int zmq::ip_addr_t::family () const
+{
+    return generic.sa_family;
+}
+
+bool zmq::ip_addr_t::is_multicast () const
+{
+    if (family () == AF_INET) {
+        uint32_t addr = ntohl (ipv4.sin_addr.s_addr);
+
+        //  IPv4 Multicast: address MSBs are 1110
+        //  Range: 224.0.0.0 - 239.255.255.255
+        return (addr & 0xf0000000) == 0xe0000000;
+    } else {
+        const uint8_t *addr = (const uint8_t *) &ipv6.sin6_addr;
+
+        //  IPv6 Multicast: ff00::/8
+        return addr[0] == 0xff;
+    }
+}
+
+uint16_t zmq::ip_addr_t::port () const
+{
+    if (family () == AF_INET6) {
+        return ntohs (ipv6.sin6_port);
+    } else {
+        return ntohs (ipv4.sin_port);
+    }
+}
+
+void zmq::ip_addr_t::set_port (uint16_t port)
+{
+    if (family () == AF_INET6) {
+        ipv6.sin6_port = htons (port);
+    } else {
+        ipv4.sin_port = htons (port);
+    }
+}
+
+//  Construct an "ANY" address for the given family
+zmq::ip_addr_t zmq::ip_addr_t::any (int family)
+{
+    ip_addr_t addr;
+
+    if (family == AF_INET) {
+        sockaddr_in *ip4_addr = &addr.ipv4;
+        memset (ip4_addr, 0, sizeof (*ip4_addr));
+        ip4_addr->sin_family = AF_INET;
+        ip4_addr->sin_addr.s_addr = htonl (INADDR_ANY);
+    } else if (family == AF_INET6) {
+        sockaddr_in6 *ip6_addr = &addr.ipv6;
+
+        memset (ip6_addr, 0, sizeof (*ip6_addr));
```

Comment:
Standard library has in6addr_any that can be assigned to sin6_addr, again if it works on Windows

Reply:
ah never mind it's 2 lines below, duh

Reply:
This is code I lifted directly from the original TCP resolver implementation:

https://github.com/zeromq/libzmq/blob/d062edd8c142384792955796329baf1e5a3377cd/src/tcp_address.cpp#L417-L427

I don't know if there's a good reason for not using the standard constants in this case. I can try I guess.

Reply:
nah it's fine as is, in6addr_any is the standard constant

Reply:
So much for that:
```
../src/ip_resolver.cpp: In static member function ‘static zmq::ip_addr_t zmq::ip_addr_t::any(int)’:
../src/ip_resolver.cpp:73:73: error: extended initializer lists only available with -std=c++0x or -std=gnu++0x [-Werror]
../src/ip_resolver.cpp:73:73: error: extended initializer lists only available with -std=c++0x or -std=gnu++0x [-Werror]
```

Reply:
ok, testing a build on a bunch of distros&archs here: https://build.opensuse.org/package/show/home:bluca:branches:network:messaging:zeromq:git-draft/libzmq

Reply:
never mind then! thanks for trying